[Model] 1.58bits BitNet Model Support

Alex4210987 opened the pull request:
This pull request is a follow-up to PR https://github.com/vllm-project/vllm/pull/7725. We fix the bug in it.

Alex4210987 committed:
[Bugfix] Fix merge error and ci error

Signed-off-by: xinyuxiao <xinyuxiao2024@gmail.com>

Alex4210987 committed:
config bugfix

Signed-off-by: xinyuxiao <xinyuxiao2024@gmail.com>

Alex4210987 committed:
Merge remote-tracking branch 'upstream/main' into bitblas-intg

Signed-off-by: xinyuxiao <xinyuxiao2024@gmail.com>

Alex4210987 committed:
config bugfix

Signed-off-by: xinyuxiao <xinyuxiao2024@gmail.com>

Alex4210987 committed:
linear.py bugfix

Signed-off-by: xinyuxiao <xinyuxiao2024@gmail.com>

Alex4210987 committed:
parameter.py bugfix

Signed-off-by: xinyuxiao <xinyuxiao2024@gmail.com>

Alex4210987 committed:
format bugfix

Signed-off-by: xinyuxiao <xinyuxiao2024@gmail.com>

Alex4210987 committed:
fix test.txt

Signed-off-by: xinyuxiao <xinyuxiao2024@gmail.com>

Alex4210987 committed:
fix test.txt

Signed-off-by: xinyuxiao <xinyuxiao2024@gmail.com>

Alex4210987 committed:
fix test.txt

Signed-off-by: xinyuxiao <xinyuxiao2024@gmail.com>

Alex4210987 committed:
fix test.txt

Signed-off-by: xinyuxiao <xinyuxiao2024@gmail.com>

Alex4210987 committed:
fix test.in

Signed-off-by: xinyuxiao <xinyuxiao2024@gmail.com>

Alex4210987 committed:
bugfix quant config

Signed-off-by: xinyuxiao <xinyuxiao2024@gmail.com>

Alex4210987 committed:
format fix

Signed-off-by: xinyuxiao <xinyuxiao2024@gmail.com>

Alex4210987 committed:
config fix

Signed-off-by: xinyuxiao <xinyuxiao2024@gmail.com>

Alex4210987 committed:
config fix

Signed-off-by: xinyuxiao <xinyuxiao2024@gmail.com>

Alex4210987 committed:
config fix

Signed-off-by: xinyuxiao <xinyuxiao2024@gmail.com>

Alex4210987 committed:
config fix

Signed-off-by: xinyuxiao <xinyuxiao2024@gmail.com>

Alex4210987 committed:
add bitnet

Signed-off-by: xinyuxiao <xinyuxiao2024@gmail.com>

Alex4210987 committed:
fix bitnet

Signed-off-by: xinyuxiao <xinyuxiao2024@gmail.com>

Alex4210987 committed:
format fix

Signed-off-by: xinyuxiao <xinyuxiao2024@gmail.com>

Alex4210987 committed:
format fix

Signed-off-by: xinyuxiao <xinyuxiao2024@gmail.com>

Alex4210987 committed:
Merge remote-tracking branch 'upstream/main' into bitblas-intg

Signed-off-by: xinyuxiao <xinyuxiao2024@gmail.com>

Alex4210987 committed:
Remove extra files

Signed-off-by: xinyuxiao <xinyuxiao2024@gmail.com>

Alex4210987 committed:
Format check

Signed-off-by: xinyuxiao <xinyuxiao2024@gmail.com>

Alex4210987 commented:
@mgoin maybe we can do a full CI check for the bitnet implementation?

Alex4210987 committed:
bugfix

Signed-off-by: xinyuxiao <xinyuxiao2024@gmail.com>

Alex4210987 committed:
bugfix

Signed-off-by: xinyuxiao <xinyuxiao2024@gmail.com>

nivibilla commented:
Would this pr also support the new falcon bitnet models?
https://huggingface.co/tiiuae/Falcon-E-3B-Instruct?s=09

younesbelkada commented:
Very interested in this work ! Happy to support if you need any help to make it work with Falcon-E models

Alex4210987 commented:
> Would this pr also support the new falcon bitnet models? https://huggingface.co/tiiuae/Falcon-E-3B-Instruct?s=09

@LeiWang1999 Maybe you can take a look?

younesbelkada commented:
Hi everyone - what is the current status of the integration ? Happy to help / take a look if there is any blocker

younesbelkada commented:
For old bitnet models, e.g.: https://huggingface.co/tiiuae/Falcon-E-3B-Instruct?s=09 per my understanding the process should be to replace linear layers with the ones from bitblas as the architecture is llama right ? 
Would it makes sense to support general bitnet models (previous design) in a separate PR?

Commit:
Merge remote-tracking branch 'upstream/main' into bitblas-intg

Signed-off-by: Your Name <you@example.com>

Commit:
Add is_init_field function to check if dataclass fields are initialized

This new utility function verifies if a specified field in a dataclass has been initialized, raising appropriate errors for non-dataclass types or missing fields. Additionally, minor formatting adjustments were made in the VllmConfig class for consistency.

Signed-off-by: Your Name <you@example.com>

aarnphm commented:
> Would it makes sense to support general bitnet models (previous design) in a separate PR?

We should do it in separate PR, given that this is already pretty large. cc @Alex4210987 wrt progress on this.

Commit:
fix lint

Signed-off-by: Your Name <you@example.com>

Commit:
fix lint

Signed-off-by: Your Name <you@example.com>

Commit:
fix lint

Signed-off-by: Your Name <you@example.com>

Commit:
fix lint

Signed-off-by: Your Name <you@example.com>

Commit:
fix lint

Signed-off-by: Your Name <you@example.com>

Commit:
fix lint

Signed-off-by: Your Name <you@example.com>

Commit:
fix lint

Signed-off-by: Your Name <you@example.com>

Commit:
fix lint

Signed-off-by: Your Name <you@example.com>

Commit:
fix lint

Signed-off-by: Your Name <you@example.com>

Commit:
fix import

Signed-off-by: Your Name <you@example.com>

aarnphm commented on benchmarks/kernels/benchmark_bitnet.py:
```diff
@@ -0,0 +1,667 @@
+# SPDX-License-Identifier: Apache-2.0
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
```
Hmm, there should be a `SPDX-FileCopyrightText:`. You can try to do `pre-commit run -a` to add this.

Alex4210987 replied:
Okay, I've added it.

aarnphm commented on tests/models/test_gptq_bitblas.py:
```diff
@@ -0,0 +1,61 @@
+# SPDX-License-Identifier: Apache-2.0
```
If these tests are light enough (which seems like it), let's also add it to `.buildkite`.

aarnphm reviewed (approved):
Let's also update the documentation section for model supported with this generations of models.

I'm not too familiar with this generation of models. Can you provide with a quick command to test this out?

cc @younesbelkada to help reviewing the model + tokenizer code.

Commit:
fix lint

Signed-off-by: Your Name <you@example.com>

Commit:
add to test

Signed-off-by: Your Name <you@example.com>

Alex4210987 commented:
@aarnphm Maybe we can do a full ci check?

Commit:
fix lint

Signed-off-by: Your Name <you@example.com>

Commit:
fix dependency

Signed-off-by: Your Name <you@example.com>

Commit:
fix dependency

Signed-off-by: Your Name <you@example.com>

Commit:
Merge remote-tracking branch 'upstream/main' into bitblas-intg

Signed-off-by: Your Name <you@example.com>

Commit:
fix test

Signed-off-by: Your Name <you@example.com>
